common/scala/src/main/scala/org/apache/openwhisk/core/database/azblob/AzureBlobAttachmentStore.scala

```diff
@@ -0,0 +1,223 @@
+    val start =
+      transid.started(this, DATABASE_ATT_SAVE, s"[ATT_PUT] uploading attachment '$name' of document 'id: $docId'")
+    val blobClient = getBlobClient(docId, name)
+
+    //TODO Use BlobAsyncClient#upload(Flux<ByteBuffer>, com.azure.storage.blob.models.ParallelTransferOptions, boolean)
+    val uploadSink = Sink.fold[ByteStringBuilder, ByteString](new ByteStringBuilder)((builder, b) => builder ++= b)
+
+    val f = docStream.runWith(combinedSink(uploadSink))
+    val g = f.flatMap { r =>
+      val buff = r.uploadResult.result().compact
+      val uf = blobClient.upload(Flux.fromArray(Array(buff.asByteBuffer)), buff.size).toFuture.toScala
+      uf.map(_ => AttachResult(r.digest, r.length))
+    }
+
+    g.foreach(_ =>
+      transid
+        .finished(this, start, s"[ATT_PUT] '$prefix' completed uploading attachment '$name' of document 'id: $docId'"))
+
+    reportFailure(
+      g,
+      start,
+      failure => s"[ATT_PUT] '$prefix' internal error, name: '$name', doc: '$docId', failure: '${failure.getMessage}'")
+  }
+
+  override protected[core] def readAttachment[T](docId: DocId, name: String, sink: Sink[ByteString, Future[T]])(
+    implicit transid: TransactionId): Future[T] = {
+    require(name != null, "name undefined")
+    val start =
+      transid.started(
+        this,
+        DATABASE_ATT_GET,
+        s"[ATT_GET] '$prefix' finding attachment '$name' of document 'id: $docId'")
+    val blobClient = getBlobClient(docId, name)
+    val f = blobClient.exists().toFuture.toScala.flatMap { exists =>
+      if (exists) {
+        val bbFlux = blobClient.download()
+        val rf = Source.fromPublisher(bbFlux).map(ByteString(_)).runWith(sink)
+        rf.andThen {
+          case Success(_) =>
+            transid
+              .finished(
+                this,
+                start,
+                s"[ATT_GET] '$prefix' completed: found attachment '$name' of document 'id: $docId'")
+        }
+      } else {
+        transid
+          .finished(
+            this,
+            start,
+            s"[ATT_GET] '$prefix', retrieving attachment '$name' of document 'id: $docId'; not found.",
+            logLevel = Logging.ErrorLevel)
+        Future.failed(NoDocumentException("Not found on 'readAttachment'."))
+      }
+    }
+
+    reportFailure(
+      f,
+      start,
+      failure =>
+        s"[ATT_GET] '$prefix' internal error, name: '$name', doc: 'id: $docId', failure: '${failure.getMessage}'")
+  }
+
+  override protected[core] def deleteAttachments(docId: DocId)(implicit transid: TransactionId): Future[Boolean] = {
+    val start =
+      transid.started(
+        this,
+        DATABASE_ATTS_DELETE,
+        s"[ATT_DELETE] deleting attachments of document 'id: $docId' with prefix ${objectKeyPrefix(docId)}")
+
+    val f = Source
+      .fromPublisher(client.listBlobsByHierarchy(objectKeyPrefix(docId)))
+      .mapAsync(1)(b => client.getBlobAsyncClient(b.getName).delete().toFuture.toScala)
```

Comment:
We've seen cases where this hierarchy delete takes a long time - it would be good to add some more logging here to indicate a) how long the listBlobsByHierarchy takes b) how many attachments were deleted, including if 0 are deleted, and c) how long each separate delete takes.